src/base/TypeUtil.ml

```diff
@@ -443,6 +443,28 @@ module TypeUtilities = struct
   let rec map_depth mt =
     match mt with MapType (_, vt) -> 1 + map_depth vt | _ -> 0
 
+  let address_field_type f t =
+    match t with
+    | Address fts ->
+        if String.(get_id f = ContractUtil.balance_label)
+        then
+          pure ContractUtil.balance_typ
+        else
+          let loc_removed = List.map fts ~f:(fun (f, t) -> (get_id f, t)) in
+          (match List.Assoc.find loc_removed (get_id f)
+                   ~equal:String.( = ) with
```

Comment:
I think we can do away with `loc_removed` and just use `equal_id` inside a `List.find`

Reply:
Unfortunately, `List.Assoc.find` requires that the `equal` argument has type `'a -> 'a -> bool`, which for us means that the annotation on identifiers needs to be the same both for the read and the address field type. The type `typ` is declared outside of the functor menagerie, so its annotation is always just `loc`, but the identifier of the reads is updated for each phase of the checker.

If you want me to use `equal_id` I will need to write my own version of `find` where the `equal` argument has type `'a -> bool`. It's not difficult to do, but we can't use the version from `List.Assoc`.

Reply:
No no, don't bother.